mysql操作频繁情况下，报错java.sql.SQLException: No operations allowed after statement closed.

### Search before asking

- [X] I had searched in the [issues](https://github.com/DTStack/chunjun/issues) and found no similar question.

- [ ] I had googled my question but i didn't get any help.

- [ ] I had read the documentation: [ChunJun doc](https://dtstack.github.io/chunjun) but it didn't help me.


### Description

我的应用场景
   emqx与redis进行联合查询，然后把处理的数据更新到mysql；

一开始对数据库更新都没有问题，但是运行不到10分钟就会报错
![image](https://user-images.githubusercontent.com/49744386/181695573-b62f22de-e8d3-46ba-8caf-0a06fd161554.png)
报错信息如下
2022-07-29 14:18:59
org.apache.flink.runtime.JobException: Recovery is suppressed by NoRestartBackoffTimeStrategy
	at org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler.handleFailure(ExecutionFailureHandler.java:118)
	at org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler.getFailureHandlingResult(ExecutionFailureHandler.java:80)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.handleTaskFailure(DefaultScheduler.java:233)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.maybeHandleTaskFailure(DefaultScheduler.java:224)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.updateTaskExecutionStateInternal(DefaultScheduler.java:215)
	at org.apache.flink.runtime.scheduler.SchedulerBase.updateTaskExecutionState(SchedulerBase.java:666)
	at org.apache.flink.runtime.scheduler.SchedulerNG.updateTaskExecutionState(SchedulerNG.java:89)
	at org.apache.flink.runtime.jobmaster.JobMaster.updateTaskExecutionState(JobMaster.java:446)
	at sun.reflect.GeneratedMethodAccessor88.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcInvocation(AkkaRpcActor.java:305)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcMessage(AkkaRpcActor.java:212)
	at org.apache.flink.runtime.rpc.akka.FencedAkkaRpcActor.handleRpcMessage(FencedAkkaRpcActor.java:77)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleMessage(AkkaRpcActor.java:158)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor$$Lambda$104/964061023.apply(Unknown Source)
	at akka.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:26)
	at akka.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:21)
	at scala.PartialFunction$class.applyOrElse(PartialFunction.scala:123)
	at akka.japi.pf.UnitCaseStatement.applyOrElse(CaseStatements.scala:21)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:170)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:517)
	at akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:225)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:592)
	at akka.actor.ActorCell.invoke(ActorCell.scala:561)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.RuntimeException: Connection maybe closed
	at com.dtstack.chunjun.connector.jdbc.sink.JdbcOutputFormat.processWriteException(JdbcOutputFormat.java:353)
	at com.dtstack.chunjun.connector.jdbc.sink.JdbcOutputFormat.writeSingleRecordInternal(JdbcOutputFormat.java:197)
	at com.dtstack.chunjun.sink.format.BaseRichOutputFormat.writeSingleRecord(BaseRichOutputFormat.java:466)
	at com.dtstack.chunjun.sink.format.BaseRichOutputFormat.lambda$writeRecordInternal$1(BaseRichOutputFormat.java:487)
	at com.dtstack.chunjun.sink.format.BaseRichOutputFormat$$Lambda$1349/844692465.accept(Unknown Source)
	at java.util.ArrayList.forEach(ArrayList.java:1249)
	at com.dtstack.chunjun.sink.format.BaseRichOutputFormat.writeRecordInternal(BaseRichOutputFormat.java:487)
	at com.dtstack.chunjun.sink.format.BaseRichOutputFormat.lambda$initTimingSubmitTask$0(BaseRichOutputFormat.java:443)
	at com.dtstack.chunjun.sink.format.BaseRichOutputFormat$$Lambda$1264/107466768.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLException: No operations allowed after statement closed.
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:965)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:898)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:887)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:861)
	at com.mysql.jdbc.StatementImpl.checkClosed(StatementImpl.java:442)
	at com.mysql.jdbc.ServerPreparedStatement.checkClosed(ServerPreparedStatement.java:510)
	at com.mysql.jdbc.ServerPreparedStatement.setString(ServerPreparedStatement.java:1971)
	at com.dtstack.chunjun.connector.jdbc.statement.FieldNamedPreparedStatementImpl.setString(FieldNamedPreparedStatementImpl.java:207)
	at com.dtstack.chunjun.connector.jdbc.converter.JdbcRowConverter.lambda$createExternalConverter$a54efc5$8(JdbcRowConverter.java:201)
	at com.dtstack.chunjun.connector.jdbc.converter.JdbcRowConverter$$Lambda$1256/685383841.serialize(Unknown Source)
	at com.dtstack.chunjun.connector.jdbc.converter.JdbcRowConverter.lambda$wrapIntoNullableExternalConverter$10addbde$1(JdbcRowConverter.java:84)
	at com.dtstack.chunjun.connector.jdbc.converter.JdbcRowConverter$$Lambda$1257/697556710.serialize(Unknown Source)
	at com.dtstack.chunjun.connector.jdbc.converter.JdbcRowConverter.toExternal(JdbcRowConverter.java:113)
	at com.dtstack.chunjun.connector.jdbc.converter.JdbcRowConverter.toExternal(JdbcRowConverter.java:51)
	at com.dtstack.chunjun.connector.jdbc.sink.PreparedStmtProxy.writeSingleRecordInternal(PreparedStmtProxy.java:200)
	at com.dtstack.chunjun.connector.jdbc.sink.JdbcOutputFormat.writeSingleRecordInternal(JdbcOutputFormat.java:194)
	... 14 more








百度的结果都说，这里和数据库的连接Connection是一个[Static]所以第一次对数据库操作没问题，当把Connection关闭后，第二次还想操作数据库时Connection肯定不存在了。
希望能完善这部分代码！

![image](https://user-images.githubusercontent.com/49744386/181695791-dfddaa61-5fba-4ba4-a96a-135a0a5932a4.png)

谢谢！！


### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/DynamicPreparedStmt.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/DynamicPreparedStmt.java
@@ -182,6 +182,10 @@ public class DynamicPreparedStmt {
         }
     }
 
+    public void reOpenStatement(Connection connection) throws SQLException {
+        this.fieldNamedPreparedStatement.reOpen(connection);
+    }
+
     public void close() throws SQLException {
         fieldNamedPreparedStatement.close();
     }

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/JdbcOutputFormat.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/JdbcOutputFormat.java
@@ -28,6 +28,7 @@ import com.dtstack.chunjun.element.ColumnRowData;
 import com.dtstack.chunjun.enums.EWriteMode;
 import com.dtstack.chunjun.enums.Semantic;
 import com.dtstack.chunjun.sink.format.BaseRichOutputFormat;
+import com.dtstack.chunjun.throwable.ChunJunRuntimeException;
 import com.dtstack.chunjun.throwable.WriteRecordException;
 import com.dtstack.chunjun.util.ExceptionUtil;
 import com.dtstack.chunjun.util.GsonUtil;
@@ -37,6 +38,7 @@ import com.dtstack.chunjun.util.TableUtil;
 import org.apache.flink.table.data.GenericRowData;
 import org.apache.flink.table.data.RowData;
 import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.util.FlinkRuntimeException;
 
 import org.apache.commons.collections.CollectionUtils;
 import org.apache.commons.lang3.StringUtils;
@@ -233,6 +235,29 @@ public class JdbcOutputFormat extends BaseRichOutputFormat {
         } finally {
             // 执行完后清空batch
             stmtProxy.clearBatch();
+        }
+    }
+
+    @Override
+    public synchronized void writeRecord(RowData rowData) {
+        checkConnValid();
+        super.writeRecord(rowData);
+    }
+
+    public void checkConnValid() {
+        try {
+            LOG.debug("check db connection valid..");
+            if (!dbConn.isValid(10)) {
+                if (Semantic.EXACTLY_ONCE == semantic) {
+                    throw new FlinkRuntimeException(
+                            "jdbc connection is valid!work's semantic is ExactlyOnce.To prevent data loss,we don't try to reopen the connection");
+                }
+                LOG.info("db connection reconnect..");
+                dbConn = getConnection();
+                stmtProxy.reOpen(dbConn);
+            }
+        } catch (Exception e) {
+            throw new ChunJunRuntimeException("failed to check jdbcConnection valid", e);
         }
     }
 

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/PreparedStmtProxy.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/PreparedStmtProxy.java
@@ -50,6 +50,7 @@ import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
@@ -73,6 +74,8 @@ public class PreparedStmtProxy implements FieldNamedPreparedStatement {
     /** LUR cache key info: database_table_rowkind * */
     protected Cache<String, DynamicPreparedStmt> pstmtCache;
 
+    protected boolean cacheIsExpire = false;
+
     /** 当前的执行sql的preparestatement */
     protected transient FieldNamedPreparedStatement currentFieldNamedPstmt;
 
@@ -93,6 +96,7 @@ public class PreparedStmtProxy implements FieldNamedPreparedStatement {
         this.connection = connection;
         this.jdbcDialect = jdbcDialect;
         this.writeExtInfo = writeExtInfo;
+        this.cacheIsExpire = true;
         initCache(true);
     }
 
@@ -361,6 +365,21 @@ public class PreparedStmtProxy implements FieldNamedPreparedStatement {
     @Override
     public void close() throws SQLException {
         currentFieldNamedPstmt.close();
+    }
+
+    @Override
+    public void reOpen(Connection connection) throws SQLException {
+        this.connection = connection;
+        ConcurrentMap<String, DynamicPreparedStmt> stringDynamicPreparedStmtConcurrentMap =
+                pstmtCache.asMap();
+        initCache(cacheIsExpire);
+        for (Map.Entry<String, DynamicPreparedStmt> entry :
+                stringDynamicPreparedStmtConcurrentMap.entrySet()) {
+            DynamicPreparedStmt value = entry.getValue();
+            value.reOpenStatement(connection);
+            pstmtCache.put(entry.getKey(), value);
+            currentFieldNamedPstmt = value.getFieldNamedPreparedStatement();
+        }
     }
 
     public void clearStatementCache() {

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/statement/FieldNamedPreparedStatement.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/statement/FieldNamedPreparedStatement.java
@@ -263,4 +263,7 @@ public interface FieldNamedPreparedStatement extends AutoCloseable {
      * @see PreparedStatement#close()
      */
     void close() throws SQLException;
+
+    /** *Reopen the Statement */
+    void reOpen(Connection connection) throws SQLException;
 }

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/statement/FieldNamedPreparedStatementImpl.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/statement/FieldNamedPreparedStatementImpl.java
@@ -39,11 +39,14 @@ import static org.apache.flink.util.Preconditions.checkNotNull;
 /** Simple implementation of {@link FieldNamedPreparedStatement}. */
 public class FieldNamedPreparedStatementImpl implements FieldNamedPreparedStatement {
 
-    private final PreparedStatement statement;
+    private PreparedStatement statement;
+    private final String parsedSQL;
     private final int[][] indexMapping;
 
-    private FieldNamedPreparedStatementImpl(PreparedStatement statement, int[][] indexMapping) {
+    private FieldNamedPreparedStatementImpl(
+            PreparedStatement statement, String parsedSQL, int[][] indexMapping) {
         this.statement = statement;
+        this.parsedSQL = parsedSQL;
         this.indexMapping = indexMapping;
     }
 
@@ -71,7 +74,7 @@ public class FieldNamedPreparedStatementImpl implements FieldNamedPreparedStatem
         }
 
         return new FieldNamedPreparedStatementImpl(
-                connection.prepareStatement(parsedSQL), indexMapping);
+                connection.prepareStatement(parsedSQL), parsedSQL, indexMapping);
     }
 
     /**
@@ -269,5 +272,11 @@ public class FieldNamedPreparedStatementImpl implements FieldNamedPreparedStatem
     @Override
     public void close() throws SQLException {
         statement.close();
+    }
+
+    @Override
+    public void reOpen(Connection connection) throws SQLException {
+        statement = null;
+        statement = connection.prepareStatement(parsedSQL);
     }
 }

